doc/architecture-decisions/0012-stopped-app-routing.md

```diff
@@ -0,0 +1,66 @@
+# 12. Stopped App Routing
+
+Date: 2020-05-29
+
+## Status
+
+Proposal
+
+## Context
+
+In CF-for-VMs, when a route is mapped to multiple apps, some started and some
+stopped, gorouter only directs requests to the apps that are started.
+
+In CF-for-k8s _currently_, we balance traffic equally between the started and
+stopped apps. This results in 503 errors for the user, and is not feature
+pairity.
+
+This is because the components that create the Istio configuration do not
+currently know about the status of apps. They assume all destinations on a Route
+object are up and able to receive traffic.
+
+See this issue: https://github.com/cloudfoundry/cf-for-k8s/issues/158
+
+## Proposal
+
+We have two proposals:
+
+1. Cloud Controller only adds destinations to Routes that are running. This way
```

Comment:
I think there's a distinction to be made here between apps that are stopped intentionally, i.e. `cf stop`, and apps that are crashing or haven't come up yet or otherwise unhealthy.  Which case is this proposal trying to address?

Reply:
Stopped intentionally. I would expect a crashed or not-yet-ready app to return 503s.

Reply:
Ah, great.  Then we can rely on CAPI's state and not have to reach out to k8s or eirini to figure it out 👍

Reply:
Could just update this thing and this would be implemented. 😅 
https://github.com/cloudfoundry/cloud_controller_ng/blob/9461ff92e68382b541c76ff27f8111eb2f111cb6/lib/cloud_controller/backends/copilot_runner_delegator.rb#L2

Reply:
One interesting thing about this, though, is that the Route CRs in k8s will no longer match as closely with what `/v3/routes` return... that's probably ok. I think I just change my mental model of what "desired vs actual state" means for routes.

Reply:
> I would expect a crashed or not-yet-ready app to return 503s.
I'd expect requests to a *route* to return 503 only if all of its destinations are stopped or crashed.  If any of them are healthy, then I'd expect requests to that route to succeed.

Reply:
> the Route CRs in k8s will no longer match as closely with what`/v3/routes` return... that's probably ok

This would make me sad.  I'd really like Route to be a high-level description of desired intent, not an intermediate representation of the current started/stopped state of its destinations.

Reply:
> I'd really like Route to be a high-level description of desired intent

In that case, does proposal 2 work for you? That way the Route resource expresses routing intent to App resources which exist regardless of the Pods, instead of expressing a desire to route to Pods that may or may not exist. (imo, a desire to route to a backend that doesn't exist is most correctly served by a 503 error)

Reply:
Yeah, that's what I thought as well... but my "probably ok" was thinking in terms of hitting GA targets for this. I think doing this optional wouldn't preclude us from moving to a more pure high-level description of desired intent later (like with the `App` CRD approach and what @jrussett is getting at in [this comment](https://github.com/cloudfoundry/cf-k8s-networking/pull/46/files#r432754776)).

As an aside, I was surprised to find out Istio was behaving this way... I thought it would behave more like how `Services` do when a subset of `Endpoints` are no longer healthy and just not route to them. I had imagined the `VirtualService` would understand that some of its backends were unhealthy and ignore them... even if we did declare some percentage of traffic should go to those. After this behavior was described to me what Istio is doing sounds like a valid approach -- just suprising to me, though, based on how other things in Kubernetes work.

Reply:
Can someone convince me the same way @tcdowney 's been convinced that `Route`s and `VirtualService`s should continue trying to route to backends that don't exist? 

> a desire to route to a backend that doesn't exist is most correctly served by a 503 error

But when 1/3 backends don't exist, a desire to route to 3 backends is most correctly served by a 503 error?